### ISSUE TITLE
Bug: service name detection fails when the app is called from unmanaged code (e.g. IIS)

### ISSUE THREAD

**Issue by gregkalapos** [elastic/apm-agent-dotnet#50]
If the user does not give us a service name we automatically use the name of the entry assembly:

```
            get
            {
                if(service == null)
                {
                    service = new Service
                    {
                        Name = Assembly.GetEntryAssembly()?.GetName()?.Name,
                        Agent = new Model.Payload.Agent
                        {
                            Name = Consts.AgentName,
                            Version = Consts.AgentVersion
                        }
                    };
                }

                return service;
            }
```

This bugs, because `Assembly.GetEntryAssembly()` can return null as the [docu](
https://docs.microsoft.com/en-us/dotnet/api/system.reflection.assembly.getentryassembly?view=netframework-4.7.2) says:

> The GetEntryAssembly method can return null when a managed assembly has been loaded from an unmanaged application. 

This happens e.g. on Full Framework when the app runs in IIS.

**Comment by gregkalapos** [comment]
First idea: We could just walk up the stack and look for the first non elastic frame. This prototype code uses the assembly name, once we sign our assemblies we should compare the public key tokens, that imo would be totally reliable.

```
if(String.IsNullOrEmpty(service.Name))
{
    var stackFrames = new StackTrace().GetFrames();

    foreach (var frame in stackFrames)
    {
        var currentAssembly = frame.GetMethod()?.DeclaringType.Assembly.GetName().Name;
        if (!String.IsNullOrEmpty(currentAssembly) && !currentAssembly.StartsWith("Elastic.Apm"))
        {
            service.Name = currentAssembly;
            break;
        }
    }

    service.Name = "unknown";
}

```

Comments:

- The Service instance is only calculated once, so this would only happen once, which is acceptable from the performance perspective.
- I know that with some application architectures we may not hit the correct assembly. I'd say it's impossible to select the right one, so the behaviour is that if users don't set the service name we set it to something which is more meaningful than "unknown" or something like that, but of course it won't be always perfect. On the other hand, users are free to set it manually if they are not happy with our heuristic.

**Comment by rushtu22** [comment]
Hello, 

My application name is BL.Mines.API so Assembly.GetEntryAssembly()?.GetName()?.Name get this name, but APM throwing validation error because of dots. So please add ServiceName to the configuration file. 

Error is data validation error: Problem validating JSON document against schema: I[#] S[#] doesn't validate with "transaction#"
  I[#/service/name] S[#/properties/service/properties/name/pattern] does not match pattern "^[a-zA-Z0-9 _-]+$"

**Comment by gregkalapos** [comment]
Super useful feedback @rushtu22! Thx for that! I will take that into consideration. A setting is definitely planned (as other agents have it). Will update here once I push something.

Thx again! 
Greg

**Comment by gregkalapos** [comment]
@rushtu22 

Also, as a workaround for now, you can fix it by doing this before the first transaction in your app: 

```
Elastic.Apm.Agent.Tracer.Service = new Service
				{
					Name = "BLMinesAPI",
					Agent = new Service.AgentC
					{
						Name = "dotNet",
						Version = "0.1"
					}
				};
```

It's a little bit ugly that the `Agent` also has to be specified, so this part of the code clearly needs more love. Will fix it!

**Comment by gregkalapos** [comment]
Fixed by #67. 

Also added a config to set the service name - by default this setting should be used. We replace `.` automatically to `_` in order to prevent errors described above. This applies to both when the service name is provided as a config setting and also when it's automatically calculated. 

In case the server is started without explicitly setting the service name it first tries to use the name of the entry assembly by using `Assembly.GetEntryAssembly()`. In case of `myApp.exe` this would be `myApp`.

In case this fails we walk the stack and move up until the current frame is not from us (signed by our .snk key) or not mscorlib (we compare the publicKeyTokens). Once we reach a frame where this is the case we set the name of the assembly that contains the given method as the service name.

As described above once there is a library between the agent frames and the application frame this can potentially report a service name that is not ideal. In my opinion this is acceptable since this can be overwritten by specifically setting the service name. Additionally once we have out of the box support for e.g. ASP.NET Core classic, we can also set the autocalculated service name to e.g. the current app pool name or to something meaningful.
